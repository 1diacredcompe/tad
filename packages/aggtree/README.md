# `aggtree`

> TODO: description

## Usage

```
const aggtree = require('aggtree');

// TODO: DEMONSTRATE API
```
