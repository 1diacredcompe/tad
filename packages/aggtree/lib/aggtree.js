'use strict';

module.exports = aggtree;

function aggtree() {
    // TODO
}
