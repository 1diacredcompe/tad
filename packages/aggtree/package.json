{
  "name": "aggtree",
  "version": "0.10.0",
  "description": "pivot table built on top of reltab",
  "author": "Antony Courtney <antony.courtney@gmail.com>",
  "homepage": "https://github.com/antonycourtney/tad-pkg#readme",
  "license": "MIT",
  "main": "dist/aggtree.js",
  "types": "dist/aggtree.d.ts",
  "scripts": {
    "prepublish": "npm run build",
    "clean": "rimraf dist",
    "build": "tsc",
    "test": "jest --config jest.config.json"
  },
  "dependencies": {
    "lodash": "^4.17.15",
    "loglevel": "^1.6.8",
    "reltab": "^0.10.0",
    "reltab-sqlite": "^0.10.0"
  },
  "devDependencies": {
    "@types/jest": "^25.2.1",
    "@types/node": "^13.11.0",
    "babel-plugin-flow-to-typescript": "^0.6.0",
    "console.table": "^0.10.0",
    "jest": "^25.2.7",
    "mkdirp": "^1.0.4",
    "prettier": "^2.0.2",
    "ts-jest": "^25.3.1",
    "typescript": "^3.8.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/antonycourtney/tad-pkg.git"
  },
  "bugs": {
    "url": "https://github.com/antonycourtney/tad-pkg/issues"
  }
}
