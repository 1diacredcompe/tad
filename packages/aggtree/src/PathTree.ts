import * as _ from "lodash";
/**
 * compact representation of a set of open paths encoded in an object
 * using keys as path components.
 * For example:
 * { "Executive Management": { "General Manager": {} }, "Safety": {},  }
 * represents the two paths:
 *    'Executive Management'/'General Manager'
 *     'Executive Management'/'Safety'
 *
 * We explicitly include null in Path elements to account for null values
 * in underlying data set.
 */
export type Path = Array<string | null>;

/*
 * N.B.: components of NodeMap have type string, not ?string;
 * We use JSON encoding / parsing of path components to ensure
 * correct handling of null
 */

type NodeMap = {
  [elem: string]: NodeMap;
}; // internal rep, not exported

const mkPathObj = (path: Path, idx: number = 0): NodeMap => {
  let ret: NodeMap = {};

  if (idx === path.length - 1) {
    ret[JSON.stringify(path[idx])] = {};
  } else {
    if (idx < path.length - 1) {
      ret[JSON.stringify(path[idx])] = mkPathObj(path, idx + 1);
    }
  }

  return ret;
};

const extendNodeMap = (nm: NodeMap, path: Path): NodeMap => {
  if (path.length === 0) {
    return nm;
  }

  const head = JSON.stringify(path[0]);
  const rest = path.slice(1);
  const subMap = nm[head];
  const restMap = subMap ? extendNodeMap(subMap, rest) : mkPathObj(rest);
  const headObj: NodeMap = {};
  headObj[head] = restMap;
  return _.defaults(nm, headObj);
}; // remove path from node map:

const removeNodeMap = (nodeMap: NodeMap, path: Path): NodeMap => {
  if (!nodeMap) {
    return {};
  }

  const head = JSON.stringify(path[0]);
  const rest = path.slice(1);

  if (rest.length === 0) {
    return _.omit(nodeMap, [head]);
  } else {
    let subMap = nodeMap[head];

    if (subMap) {
      const nextSub = removeNodeMap(subMap, rest); // immutable replacement of nodeMap[head]:

      const ret = Object.assign({}, nodeMap);
      ret[head] = nextSub;
      return ret;
    } else {
      // head not in nodeMap?
      return nodeMap;
    }
  }
};

// trim an open node map to given depth:
const trimNodeMapToDepth = (nodeMap: NodeMap, depth: number): NodeMap => {
  if (depth === 0) {
    return {};
  }

  let ret: NodeMap = {};

  for (let elem in nodeMap) {
    ret[elem] = trimNodeMapToDepth(nodeMap[elem], depth - 1);
  }

  return ret;
};

function* walkNodeMap(
  prefix: Path,
  nodeMap: NodeMap
): Generator<Path, void, void> {
  // eslint-disable-line
  for (var component in nodeMap) {
    if (nodeMap.hasOwnProperty(component)) {
      let subPath = prefix.slice();
      subPath.push(JSON.parse(component));
      yield subPath; // parents before children

      let cval = nodeMap[component];

      if (typeof cval === "object") {
        yield* walkNodeMap(subPath, cval);
      }
    }
  }
}
/*
 * An immutable PathTree
 */

export class PathTree {
  _rep: NodeMap;

  constructor(nodeMap: NodeMap = {}) {
    this._rep = nodeMap;
  }

  trimToDepth(depth: number): PathTree {
    return new PathTree(trimNodeMapToDepth(this._rep, depth));
  } // extend PathTree with the given path:

  open(path: Path): PathTree {
    return new PathTree(extendNodeMap(this._rep, path));
  } // remove the given path from PathTree:

  close(path: Path): PathTree {
    const nextMap = removeNodeMap(this._rep, path);
    return new PathTree(nextMap);
  }

  isOpen(path: Path): boolean {
    let nm = this._rep;

    for (let elem of path) {
      let encElem = JSON.stringify(elem);
      nm = nm[encElem];

      if (!nm) {
        return false;
      }
    }

    return true;
  }

  *iter(): Generator<Path, void, void> {
    yield* walkNodeMap([], this._rep);
  }
}
