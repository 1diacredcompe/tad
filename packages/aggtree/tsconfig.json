{
  "compilerOptions": {
    "outDir": "./dist",
    "downlevelIteration": true,
    "declaration": true,
    "strict": true,
    "target": "ES5",
    "lib": ["es6", "es2017", "dom"]
  },
  "include": ["./src/*", "./test/*"]
}
