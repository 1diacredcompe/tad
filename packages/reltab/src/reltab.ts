export * from "./dialectRegistry";
export * from "./Connection";
export * from "./defs";
export * from "./FilterExp";
export * from "./QueryExp";
export * from "./Schema";
export * from "./TableRep";
export * from "./AggFn";
export * from "./ColumnType";
export * from "./DataSource";
