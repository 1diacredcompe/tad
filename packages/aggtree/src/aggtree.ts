import * as reltab from "reltab";
import * as _ from "lodash";
import { Path, PathTree } from "./PathTree";
import {
  Connection,
  QueryExp,
  Schema,
  AggColSpec,
  asString,
  ValExp,
} from "reltab"; // eslint-disable-line

export * from "./PathTree";

const { col, constVal } = reltab;
const PATHSEP = "#";
const ENCPATHSEP = "%23";
/*
 *  We used to use encodeURIComponent, but this isn't readily available on
 *  SQLite and we need to use the same encoding in both places so that
 *  path sort order works out correctly.
 *
 *  We'll do this much simpler string encoding that escapes % chars and PATHSEP.
 *  Can still be decoded with decodeURIComponent.
 */

const simpleStringEncode = (str?: string | null): string | undefined | null => {
  if (str == null) {
    return null;
  }

  return str.replace("%", "%25").replace(PATHSEP, ENCPATHSEP);
};

export const encodePath = (path: Path): string => {
  // const eps = path.map(encodeURIComponent)
  const eps = path.map(simpleStringEncode);
  const ret = PATHSEP + eps.join(PATHSEP);
  return ret;
};
export const decodePath = (pathStr: string): Path => {
  pathStr = pathStr.slice(1); // drop leading PATHSEP

  const eps = pathStr.length > 0 ? pathStr.split(PATHSEP) : [];
  const path = eps.map(decodeURIComponent);
  return path;
};

const addPathCols = (
  baseQuery: QueryExp,
  baseDepth: number,
  maxDepth: number
): QueryExp => {
  let retQuery = baseQuery;

  for (let i = baseDepth; i < maxDepth - 1; i++) {
    retQuery = retQuery.extend("_path" + i, "text", asString(constVal(null)));
  }

  return retQuery;
};

export class VPivotTree {
  rt: reltab.Connection;
  baseQuery: QueryExp;
  pivotColumns: Array<string>;
  pivotLeafColumn: string | undefined | null;
  baseSchema: reltab.Schema;
  outCols: Array<string>;
  rootQuery: QueryExp | undefined | null;
  sortKey: Array<[string, boolean]>;
  aggMap:
    | {
        [cid: string]: reltab.AggFn;
      }
    | undefined
    | null;

  constructor(
    rt: Connection,
    baseQuery: QueryExp,
    baseSchema: reltab.Schema,
    pivotColumns: Array<string>,
    pivotLeafColumn: string | undefined | null,
    outCols: Array<string>,
    rootQuery: QueryExp | undefined | null,
    sortKey: Array<[string, boolean]>,
    inAggMap?: {
      [cid: string]: reltab.AggFn;
    } | null
  ) {
    this.rt = rt;
    this.pivotColumns = pivotColumns;
    this.pivotLeafColumn = pivotLeafColumn;
    this.baseQuery = baseQuery;
    this.baseSchema = baseSchema;
    this.outCols = outCols;
    this.rootQuery = rootQuery;
    this.sortKey = sortKey;
    this.aggMap = inAggMap;
  }
  /*
   * returns a query for the children of the specified path:
   */

  applyPath(path: Path): QueryExp {
    // TODO: Think about how to use rollupBy or a smaller number of groupBys that get chopped up
    // and cache result for efficiency
    // queries are immutable so no need to clone:
    var pathQuery = this.baseQuery; // recCountQuery
    // We will filter by all path components, and then group by the next pivot Column:

    if (path.length > this.pivotColumns.length) {
      throw new Error("applyPath: path length > pivot columns");
    }

    if (path.length > 0) {
      var pred = reltab.and();

      for (var i = 0; i < path.length; i++) {
        let pathElem = path[i];
        let pivotColExp = col(this.pivotColumns[i]);

        if (pathElem == null) {
          pred = pred.isNull(pivotColExp);
        } else {
          pred = pred.eq(pivotColExp, constVal(pathElem));
        }
      }

      pathQuery = pathQuery.filter(pred);
    }

    const pivotColumnInfo: reltab.ColumnMapInfo = {
      id: "_pivot",
      type: "text",
      displayName: "_pivot",
    };
    const aggCols = this.baseSchema.columns;
    const aggMap = this.aggMap;
    const gbAggs: any =
      aggMap != null ? aggCols.map((cid) => [aggMap[cid], cid]) : aggCols;

    if (path.length < this.pivotColumns.length) {
      /* was: 
      pathQuery = pathQuery
        .groupBy([this.pivotColumns[path.length]], gbAggs)
        .mapColumnsByIndex({
          "0": pivotColumnInfo,
        });
        ...but this leads to ambiguity in the use of the pivot column name, that some SQL engines (BigQuery) don't like,
        so we'll push the definition of _pivot column inside the GroupBy:
      */
      pathQuery = pathQuery
        .extend(
          "_pivot",
          "text",
          asString(col(this.pivotColumns[path.length])),
          { displayName: "_pivot" }
        )
        .groupBy(["_pivot"], gbAggs);
    } else {
      // leaf level
      const leafExp =
        this.pivotLeafColumn == null
          ? constVal("")
          : asString(col(this.pivotLeafColumn!));
      pathQuery = pathQuery.extend("_pivot", "text", leafExp);
    }

    const depth = path.length + 1;
    pathQuery = pathQuery
      .extend("_depth", "integer", constVal(depth))
      .extend("_isRoot", "boolean", constVal(0))
      .project(this.outCols);
    /*
     * The point of the '_sortVal_<i>' column is that it will be 1 for all rows of
     * depth >= i, 0 for rows where depth < i (which are higher in the pivot tree).
     * We do an ascending sort on this before
     * adding the '_sortVal_<i>_<j>' cols to ensure that parents always come
     * before their children; without this we'd end up putting parent row
     * immediately after children when sorted descending by some column
     */

    const maxDepth = this.pivotColumns.length + 1;

    for (let i = 0; i < maxDepth; i++) {
      const depthVal = depth > i ? 1 : 0;
      pathQuery = pathQuery.extend(
        "_sortVal_" + i,
        "integer",
        constVal(depthVal)
      );
    }

    for (let i = 0; i < this.pivotColumns.length; i++) {
      let pathElemExp: ValExp = constVal(null);

      if (i < path.length) {
        let colType = this.baseSchema.columnType(this.pivotColumns[i]);
        pathElemExp = constVal(path[i]);
      } else if (i === path.length) {
        pathElemExp = col("_pivot"); // SQL expression referring to _pivot column
      }

      pathQuery = pathQuery.extend("_path" + i, "text", asString(pathElemExp));
    }

    // TODO: Should we optionally also insert _childCount and _leafCount ?
    // _childCount would count next level of groupBy,
    // _leafCount would do count() at point of calculating
    // filter for current path (before doing groupBy).
    // These can certainly have non-trivial costs to calculate

    return pathQuery;
  }
  /*
   * get query for joining with pathQuery to sort to specified depth
   */

  getSortQuery(depth: number): QueryExp {
    let sortQuery = this.baseQuery; // recCountQuery

    const sortCols = this.sortKey.map((p) => p[0]);
    const aggMap = this.aggMap;
    const sortColAggs: any =
      aggMap != null ? sortCols.map((cid) => [aggMap[cid], cid]) : sortCols;
    const gbCols = this.pivotColumns.slice(0, depth);
    sortQuery = sortQuery.groupBy(gbCols, sortColAggs);
    let colMap: { [cid: string]: { id: string } } = {};

    for (let i = 0; i < gbCols.length; i++) {
      const pathColName = "_path" + i;
      colMap[gbCols[i]] = {
        id: pathColName,
      };
    }

    sortQuery = sortQuery.mapColumns(colMap);
    const pathLevel = depth - 1;
    let sortColMap: { [cid: string]: { id: string } } = {};

    for (let i = 0; i < sortCols.length; i++) {
      let colIndex = gbCols.length + i;
      let colName = "_sortVal_" + pathLevel.toString() + "_" + i.toString();
      sortColMap[colIndex.toString()] = {
        id: colName,
      };
    }

    sortQuery = sortQuery.mapColumnsByIndex(sortColMap);
    return sortQuery;
  }
  /*
   * get query for full tree state from a set of openPaths
   */

  getTreeQuery(openPaths: PathTree): QueryExp {
    const maxDepth = this.pivotColumns.length + 1;
    let resQuery = null;

    if (this.rootQuery) {
      resQuery = this.rootQuery;

      for (let i = 0; i < maxDepth; i++) {
        resQuery = resQuery.extend("_sortVal_" + i, "integer", constVal(0));
      }

      resQuery = addPathCols(resQuery, 0, maxDepth);
    }

    const openRoot = this.applyPath([]); // immediate children of root

    if (resQuery) {
      resQuery = resQuery.concat(openRoot);
    } else {
      resQuery = openRoot;
    }

    for (let path of openPaths.iter()) {
      let subQuery = this.applyPath(path);
      resQuery = resQuery.concat(subQuery);
    }

    const sortArg: [string, boolean][] = [];

    for (let i = 0; i < maxDepth - 1; i++) {
      sortArg.push(["_path" + i, true]);
    }

    resQuery = resQuery.sort(sortArg);
    return resQuery;
  }
  /*
   * get query for full tree state from a set of openPaths, joined with
   * relevant sort queries based on pivot depth, and with appropriate
   * order by clause
   */

  getSortedTreeQuery(openPaths: PathTree): QueryExp {
    const tq = this.getTreeQuery(openPaths);
    let jtq = tq; // add sort queries for each pivot depth and join to tree query

    for (let i = 0; i < this.pivotColumns.length; i++) {
      let depth = i + 1;
      let sq = this.getSortQuery(depth);

      let joinKey = _.range(0, depth).map((j) => "_path" + j);

      jtq = jtq.join(sq, joinKey);
    } // Now let's work out the sort key:
    // potential opt: Eliminate if root not shown

    let tsortKey: [string, boolean][] = [["_isRoot", false]];

    for (let i = 0; i < this.pivotColumns.length; i++) {
      // should be able to do a simple tsortKey.push for next line, but flow being lame
      tsortKey = tsortKey.concat([["_sortVal_" + i.toString(), true]]); // sort keys for this depth:

      let dsortKey: [string, boolean][] = _.range(
        0,
        this.sortKey.length
      ).map((j) => ["_sortVal_" + i + "_" + j, this.sortKey[j][1]]);

      tsortKey = tsortKey.concat(dsortKey); // splice in path at this depth:

      tsortKey = tsortKey.concat([["_path" + i, true]]);
    } // Add the final _sortVal_i:

    const maxDepth = this.pivotColumns.length;
    tsortKey = tsortKey.concat([["_sortVal_" + maxDepth.toString(), true]]); // Finally, add the sort key columns itself for leaf level:

    tsortKey = tsortKey.concat(this.sortKey);
    const stq = jtq.sort(tsortKey);
    return stq;
  }
}
export const getBaseSchema = (
  rt: reltab.Connection,
  baseQuery: QueryExp
): Promise<Schema> => {
  // add a count column and do the usual SQL where 1=0 trick:
  const schemaQuery = baseQuery
    .extend("Rec", "integer", constVal(1))
    .filter(reltab.and().eq(constVal(1), constVal(0)));
  const schemap = rt.evalQuery(schemaQuery);
  return schemap.then((schemaRes) => schemaRes.schema);
};
export function vpivot(
  rt: reltab.Connection,
  baseQuery: QueryExp,
  baseSchema: Schema,
  pivotColumns: Array<string>,
  pivotLeafColumn: string | undefined | null,
  showRoot: boolean,
  sortKey: Array<[string, boolean]>,
  inAggMap:
    | {
        [cid: string]: reltab.AggFn;
      }
    | undefined
    | null = null
): VPivotTree {
  const aggMap = inAggMap; // just for Flow

  baseQuery = baseQuery.extend("Rec", "integer", constVal(1));
  const hiddenCols = ["_depth", "_pivot", "_isRoot"];
  const outCols = baseSchema.columns.concat(hiddenCols);
  const gbCols = baseSchema.columns.slice();
  const gbAggs: AggColSpec[] =
    aggMap != null ? gbCols.map((cid) => [aggMap[cid], cid]) : gbCols;
  let rootQuery = null;

  if (showRoot) {
    rootQuery = baseQuery
      .groupBy([], gbAggs)
      .extend("_pivot", "text", asString(constVal(null)))
      .extend("_depth", "integer", constVal(0))
      .extend("_isRoot", "boolean", constVal(1))
      .project(outCols);
  }

  return new VPivotTree(
    rt,
    baseQuery,
    baseSchema,
    pivotColumns,
    pivotLeafColumn,
    outCols,
    rootQuery,
    sortKey,
    aggMap
  );
}
