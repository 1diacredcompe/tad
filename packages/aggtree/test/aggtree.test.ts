import * as sqlite3 from "sqlite3";
import * as reltab from "reltab";
import * as aggtree from "../src/aggtree";
import { PathTree } from "../src/PathTree";
import * as reltabSqlite from "reltab-sqlite";
import { textSpanContainsPosition, textSpanContainsTextSpan } from "typescript";
import { delimiter } from "path";
import * as log from "loglevel";
import * as util from "./testUtils";
import { executionAsyncId } from "async_hooks";
import { DbConnectionKey, getConnection } from "reltab";

log.setLevel("info");

let testCtx: reltabSqlite.SqliteContext;

const importCsv = async (db: sqlite3.Database, path: string) => {
  const md = await reltabSqlite.fastImport(db, path);
};

const pcols = ["JobFamily", "Title", "Union", "Name", "Base", "TCOE"];
const q0 = reltab.tableQuery("barttest").project(pcols);

let tree0: aggtree.VPivotTree;

beforeAll(
  async (): Promise<reltabSqlite.SqliteContext> => {
    log.setLevel("info"); // use "debug" for even more verbosity
    const showQueries = true;
    const connKey: DbConnectionKey = {
      providerName: "sqlite",
      connectionInfo: ":memory:",
    };
    const ctx = await getConnection(connKey);
  
    testCtx = ctx as reltabSqlite.SqliteContext;

    const db = testCtx.db;

    await importCsv(db, "../reltab-sqlite/test/support/sample.csv");
    await importCsv(db, "../reltab-sqlite/test/support/barttest.csv");

    const schema = await aggtree.getBaseSchema(testCtx, q0);
    log.debug("got schema: ", schema);

    tree0 = aggtree.vpivot(
      testCtx,
      q0,
      schema,
      ["JobFamily", "Title"],
      "Name",
      true,
      []
    );

    return testCtx;
  }
);

test("rootQuery Test", async () => {
  const rq0 = tree0.rootQuery;
  log.info("root query exp:\n", rq0?.toJS());

  const res0 = await testCtx.evalQuery(rq0!);
  log.info("root query result:");
  util.logTable(res0);
  expect(res0).toMatchSnapshot();
});

test("initial applyPath operations", async () => {
  const q1 = tree0.applyPath([]);
  log.info("open root query:");
  log.info(q1.toJS());
  const res1 = await testCtx.evalQuery(q1);
  util.logTable(res1);

  const expCols = [
    "JobFamily",
    "Title",
    "Union",
    "Name",
    "Base",
    "TCOE",
    "Rec",
    "_depth",
    "_pivot",
    "_isRoot",
    "_sortVal_0",
    "_sortVal_1",
    "_sortVal_2",
    "_path0",
    "_path1",
  ];

  expect(res1.schema.columns).toEqual(expCols);
  expect(res1.rowData.length).toBe(9);

  const actSum = util.columnSum(res1, "TCOE");

  expect(actSum).toBe(4691559);
});

test("open specific node", async () => {
  const q2 = tree0.applyPath(["Executive Management"]);

  // console.log("after opening node: q2: ", JSON.stringify(q2, null, 2));

  const res2 = await testCtx.evalQuery(q2);

  // console.log('after applying path ["Executive Management"]:');
  // util.logTable(res2);
  expect(res2).toMatchSnapshot();

  const q3 = tree0.applyPath(["Executive Management", "General Manager"]);
  const res3 = await testCtx.evalQuery(q3);

  // console.log("after applying path /Executive Management/General Manager:");
  // util.logTable(res3);
  expect(res3).toMatchSnapshot();

  const openPaths = new PathTree({ '"Executive Management"': {} });
  const q4 = tree0.getTreeQuery(openPaths);
  const res4 = await testCtx.evalQuery(q4);

  // console.log("after treeQuery for path /Executive Management: ");
  util.logTable(res4);
});

// Based on aggTreeTest1 from original Tad test suite:
test("basic sorted aggTree test", async () => {
  const q0 = reltab.tableQuery("barttest").project(pcols);
  const schema = await aggtree.getBaseSchema(testCtx, q0);
  const tree0 = aggtree.vpivot(
    testCtx,
    q0,
    schema,
    ["JobFamily", "Title"],
    "Name",
    true,
    [
      ["TCOE", false],
      ["Base", true],
      ["Title", true],
    ]
  );

  // console.log("vpivot initial promise resolved...");

  const sq1 = tree0.getSortQuery(1);

  const res = await testCtx.evalQuery(sq1);
  // console.log("sort query depth 1: ");
  // util.logTable(res);

  expect(res).toMatchSnapshot();

  const sq2 = tree0.getSortQuery(2);

  const res2 = await testCtx.evalQuery(sq2);
  // console.log("sort query depth 2: ");
  // util.logTable(res2);
  expect(res2).toMatchSnapshot();

  const q1 = tree0.applyPath([]);

  // console.log("got depth 1 query and sortQuery, joining...: ");
  const jq1 = q1.join(sq1, "_path0");
  const jres = await testCtx.evalQuery(jq1);

  // console.log("result of join query: ");
  // util.logTable(jres);
  expect(jres).toMatchSnapshot();
});

test("async aggTree sortedTreeQuery test", async () => {
  const q0 = reltab.tableQuery("barttest").project(pcols);
  const schema = await aggtree.getBaseSchema(testCtx, q0);
  const tree0 = aggtree.vpivot(
    testCtx,
    q0,
    schema,
    ["JobFamily", "Title"],
    "Name",
    true,
    [
      ["TCOE", false],
      ["Base", true],
      ["Title", true],
    ]
  );
  const rtc = testCtx;
  const openPaths = new PathTree({
    '"Executive Management"': { '"General Manager"': {} },
    '"Safety"': {},
  });
  const q4 = tree0.getTreeQuery(openPaths);
  const res4 = await rtc.evalQuery(q4);

  // console.log("tree query after opening paths:");
  // util.logTable(res4, { maxRows: 50 });

  const sq1 = tree0.getSortQuery(1);

  const res = await testCtx.evalQuery(sq1);
  // console.log("sort query depth 1: ");
  // util.logTable(res);

  expect(res).toMatchSnapshot();
  const sq2 = tree0.getSortQuery(2);

  const res2 = await testCtx.evalQuery(sq2);
  // console.log("sort query depth 2: ");
  // util.logTable(res2);
  expect(res2).toMatchSnapshot();

  const jq4 = q4.join(sq1, ["_path0"]).join(sq2, ["_path0", "_path1"]);
  const jres4 = await rtc.evalQuery(jq4);

  // console.log("tree query after sort joins:");
  // util.logTable(jres4, { maxRows: 50 });
  expect(jres4).toMatchSnapshot();

  const stq = tree0.getSortedTreeQuery(openPaths);

  console.log("full sorted tree query:\n", stq.toJS());

  const sres = await rtc.evalQuery(stq);

  // console.log("result of sorted tree query:");
  // util.logTable(sres, { maxRows: 50 });
  expect(sres).toMatchSnapshot();
});
