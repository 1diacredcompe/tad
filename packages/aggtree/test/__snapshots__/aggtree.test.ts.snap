// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial aggtree Test 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "Base": 212647,
      "JobFamily": "Executive Management",
      "Name": "Oversier, Paul",
      "Rec": 1,
      "TCOE": 288160,
      "Title": "Assistant General Manager Operations",
      "Union": "Non-Represented",
      "_depth": 2,
      "_isRoot": 0,
      "_path0": "Executive Management",
      "_path1": "Assistant General Manager Operations",
      "_pivot": "Assistant General Manager Operations",
      "_sortVal_0": 1,
      "_sortVal_1": 1,
      "_sortVal_2": 0,
    },
    Object {
      "Base": 206914,
      "JobFamily": "Executive Management",
      "Name": "DeVaughn, Marcia",
      "Rec": 1,
      "TCOE": 281548,
      "Title": "Deputy General Manager",
      "Union": "Non-Represented",
      "_depth": 2,
      "_isRoot": 0,
      "_path0": "Executive Management",
      "_path1": "Deputy General Manager",
      "_pivot": "Deputy General Manager",
      "_sortVal_0": 1,
      "_sortVal_1": 1,
      "_sortVal_2": 0,
    },
    Object {
      "Base": 611161,
      "JobFamily": "Executive Management",
      "Name": null,
      "Rec": 2,
      "TCOE": 819582,
      "Title": "General Manager",
      "Union": "Non-Represented",
      "_depth": 2,
      "_isRoot": 0,
      "_path0": "Executive Management",
      "_path1": "General Manager",
      "_pivot": "General Manager",
      "_sortVal_0": 1,
      "_sortVal_1": 1,
      "_sortVal_2": 0,
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "Base": 4,
      "JobFamily": 0,
      "Name": 3,
      "Rec": 6,
      "TCOE": 5,
      "Title": 1,
      "Union": 2,
      "_depth": 7,
      "_isRoot": 9,
      "_path0": 13,
      "_path1": 14,
      "_pivot": 8,
      "_sortVal_0": 10,
      "_sortVal_1": 11,
      "_sortVal_2": 12,
    },
    "columnMetadata": Object {
      "Base": Object {
        "displayName": "Base",
        "type": "integer",
      },
      "JobFamily": Object {
        "displayName": "Job Family",
        "type": "text",
      },
      "Name": Object {
        "displayName": "Name",
        "type": "text",
      },
      "Rec": Object {
        "type": "integer",
      },
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "text",
      },
      "Union": Object {
        "displayName": "Union",
        "type": "text",
      },
      "_depth": Object {
        "type": "integer",
      },
      "_isRoot": Object {
        "type": "boolean",
      },
      "_path0": Object {
        "type": "text",
      },
      "_path1": Object {
        "type": "text",
      },
      "_pivot": Object {
        "displayName": "_pivot",
        "type": "text",
      },
      "_sortVal_0": Object {
        "type": "integer",
      },
      "_sortVal_1": Object {
        "type": "integer",
      },
      "_sortVal_2": Object {
        "type": "integer",
      },
    },
    "columns": Array [
      "JobFamily",
      "Title",
      "Union",
      "Name",
      "Base",
      "TCOE",
      "Rec",
      "_depth",
      "_pivot",
      "_isRoot",
      "_sortVal_0",
      "_sortVal_1",
      "_sortVal_2",
      "_path0",
      "_path1",
    ],
  },
}
`;

exports[`initial aggtree Test 2`] = `
TableRep {
  "rowData": Array [
    Object {
      "Base": 312461,
      "JobFamily": "Executive Management",
      "Name": "Crunican, Grace",
      "Rec": 1,
      "TCOE": 399921,
      "Title": "General Manager",
      "Union": "Non-Represented",
      "_depth": 3,
      "_isRoot": 0,
      "_path0": "Executive Management",
      "_path1": "General Manager",
      "_pivot": "Crunican, Grace",
      "_sortVal_0": 1,
      "_sortVal_1": 1,
      "_sortVal_2": 1,
    },
    Object {
      "Base": 298700,
      "JobFamily": "Executive Management",
      "Name": "Dugger, Dorothy",
      "Rec": 1,
      "TCOE": 419661,
      "Title": "General Manager",
      "Union": "Non-Represented",
      "_depth": 3,
      "_isRoot": 0,
      "_path0": "Executive Management",
      "_path1": "General Manager",
      "_pivot": "Dugger, Dorothy",
      "_sortVal_0": 1,
      "_sortVal_1": 1,
      "_sortVal_2": 1,
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "Base": 4,
      "JobFamily": 0,
      "Name": 3,
      "Rec": 6,
      "TCOE": 5,
      "Title": 1,
      "Union": 2,
      "_depth": 7,
      "_isRoot": 9,
      "_path0": 13,
      "_path1": 14,
      "_pivot": 8,
      "_sortVal_0": 10,
      "_sortVal_1": 11,
      "_sortVal_2": 12,
    },
    "columnMetadata": Object {
      "Base": Object {
        "displayName": "Base",
        "type": "integer",
      },
      "JobFamily": Object {
        "displayName": "Job Family",
        "type": "text",
      },
      "Name": Object {
        "displayName": "Name",
        "type": "text",
      },
      "Rec": Object {
        "type": "integer",
      },
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "text",
      },
      "Union": Object {
        "displayName": "Union",
        "type": "text",
      },
      "_depth": Object {
        "type": "integer",
      },
      "_isRoot": Object {
        "type": "boolean",
      },
      "_path0": Object {
        "type": "text",
      },
      "_path1": Object {
        "type": "text",
      },
      "_pivot": Object {
        "type": "text",
      },
      "_sortVal_0": Object {
        "type": "integer",
      },
      "_sortVal_1": Object {
        "type": "integer",
      },
      "_sortVal_2": Object {
        "type": "integer",
      },
    },
    "columns": Array [
      "JobFamily",
      "Title",
      "Union",
      "Name",
      "Base",
      "TCOE",
      "Rec",
      "_depth",
      "_pivot",
      "_isRoot",
      "_sortVal_0",
      "_sortVal_1",
      "_sortVal_2",
      "_path0",
      "_path1",
    ],
  },
}
`;
